packages/diagram/src/layout/layoutEngine.ts

```diff
@@ -627,34 +634,154 @@ export class Layout {
+        const [targetPos, requiredPreceeding] = this.calculateAddArgPos(
+            key,
+            listEntries,
+            targetExpression.trailingArgumentExpressions.length > 0
+        );
+        const parenthesisRange = (targetExpression.metadata as ParenthesisExpressionMetadata).parenthesisRange;
+        const range = this.calculateAddArgRange(targetPos, listEntries, parenthesisRange);
+        return {
+            type: "add-arg",
+            key,
+            template: parsedTemplate,
+            range,
+            listRange: parenthesisRange,
+            isFirst: targetPos <= 0,
+            isLast: targetPos >= listEntries.length,
+            requiredPreceeding
+        };
+    }
+
+    /**
+     * Calculates the range which should be replaced when inserting an argument
+     *
+     * @param targetPos the position where the argument should be inserted
+     * @param listEntries the inner argument expressions
+     * @param parenthesisRange the range of the whole parenthesis
+     * @returns the range which should be replaced
+     */
+    private calculateAddArgRange(targetPos: number, listEntries: ListEntry[], parenthesisRange: Range): Range {
+        if (targetPos <= 0) {
+            if (listEntries.length === 0) {
+                return parenthesisRange;
+            } else {
+                return [parenthesisRange[0], listEntries[0].value.range[0]];
+            }
+        } else if (targetPos >= listEntries.length) {
+            return [listEntries.at(-1)!.value.range[1], parenthesisRange[1]];
+        } else {
+            return [listEntries[targetPos - 1].value.range[1], listEntries[targetPos].value.range[0]];
+        }
+    }
+
+    /**
+     * Calculates the position where an argument should be added.
+     * Semantics: Index where the argument should be inserted (e.g. 0 for the first)
+     *
+     * @param key the key of the argument
+     * @param listEntries the inner argument expressions
+     * @param hasTrailingArg true if the call expression has a trailing argument
+     * @returns the position where the argument should be added, and the amount of required before-inserted arguments
+     */
+    private calculateAddArgPos(
+        key: string | number,
+        listEntries: ListEntry[],
+        hasTrailingArg: boolean
+    ): [number, number | undefined] {
+        if (typeof key === "string") {
+            return [listEntries.length, undefined];
+        }
+        let foundIndexArgs = 0;
+        let targetPos = 0;
+        for (const entry of listEntries) {
+            if (entry.name == undefined) {
+                foundIndexArgs++;
+            }
+            if (foundIndexArgs >= key + 1) {
+                break;
+            }
+            targetPos++;
+        }
+        if (key + 1 > foundIndexArgs && hasTrailingArg) {
+            throw new Error("Cannot add argument after or in between trailing function expressions");
```

Comment:
imagine you have the following code
```
if (true) {
  // whatever
} {
  // whatever
}
```
This function call has 3 positional arguments
inserting sth at position -0.5 (before the first) or 0.5 (after first) is supported
however inserting sth at 1.5 (between the two trailing function arguments) or at 3 (after the second trailing function) is not supported, as in this case, only a function expression could successfully be inserted, however, the contract of this edit allows arbitrary expressions (to be precise in case of insert even multiple comma-separeted expressions)

Reply:
if you have a better error message for this, feel free to provide one

Reply:
`Only function expressions can be inserted behind other function expressions`?

Reply:
yes, except that I didn't implement that feature (as I doubt we will ever have a use case for this)
Also this would then need to be put in a different kind of edit type as this edit type allows to insert arbitrary expressions as part of its contract

Reply:
Also the problem is that this only affects trailing function expressions
you can have function expressions as regular positional args and it works perfectly fine